draft-ietf-ppm-dap.md

```diff
@@ -1554,12 +1563,18 @@ struct {
 where `prepare_resps` are the outbound prep steps computed in the previous step.
 The order MUST match `AggregationJobInitReq.prepare_inits`.
 
-The Helper responds to the Leader with HTTP status code 201 Created and a body
-consisting of the `AggregationJobResp`, with media type
-"application/dap-aggregation-job-resp".
+After receiving the response to its `AggregationJobInitReq`, the Leader makes
+requests to
+`GET {helper}/tasks/{task-id}/aggregation_jobs/{aggregation-job-id}` to check
+on the status of the aggregation job. If the aggregation job is ready, the
+Helper responds with HTTP status 200 OK and a body consisting of an
+`AggregationJobResp`, with media type "application/dap-aggregation-job-resp".
+If the aggregation job is not finished yet, the Helper responds with HTTP
+status 202 Accepted. The response MAY include a Retry-After header field to
```

Comment:
It looks like this requires constant polling from the leader end to check status of each job. Should it implement long polling and respond when there is response to return?

Reply:
If we are not doing long polling for any reason, the response should probably 200 with appropriate status response instead of 202, since we are not accepting a new request for processing.

Reply:
What do you mean by "long polling"?

Reply:
Hmm, I see long polling as the state of the spec before this change, more or less. That is, leaders hold the HTTP connection open while the helper does processing. The helper returns a response when it's ready.

> If we are not doing long polling for any reason, the response should probably 200 with appropriate status response instead of 202, since we are not accepting a new request for processing.

202 signals to the leader that polling needs to continue. We could give 200 with an empty body, but I think 202 is most appropriate, [given my reading of MDN](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/202). (FWIW I copied the semantics from collection, so if we argue against using 202 we would need to fix collection too).

Reply:
Earlier model is, you wait when you init a job. Now, we submit the job and then wait on a different thread. But leader has to wait for several jobs to complete in this model, right? So, it is not clear to me how leader keep polling for all jobs better than long polling.

@cjpatton - http long polling - in this context, it translates to helper blocks the GET call until the job processing is done (with an optional timeout).

Reply:
The goal is to enable the helper to move expensive computation (checking uniqueness indices, heavyweight VDAF math, etc.) out of the request hot-path. 

It is plausible to make no changes to the spec, which IIUC is equivalent to using a long-polling strategy. Implementations could still decouple expensive operations by leaving the connection open and internally polling a pool of compute workers.

But I (if I'm operating as a helper) am still left with the disadvantages of having to deal with long-running connections. It's tricky to load balance them, e.g. what do we do with existing connections when we need to scale up? Whereas for stateless GET requests with no body, load balancing them is trivial.

If we really want to avoid polling, we may want to look into the more robust 2-way protocols, i.e. websockets.

Reply:
To summarize my proposed changes - 
1. aggregate init request can return either 200 with proper response body (sync) or 202 without body (async). This makes it backward compatible and supports sync model.
2. When leader is ready to receive the init job response, it makes a GET <jobid> http request. Helper can
  * respond with 202 immediately if the job init is not done on helper (frequent polling model)
  * respond with 200 immediately with init job response (if job is completed)
  * Block the caller and respond with 200 with init job response when job is finished (long polling).

Reply:
I agree that the old semantics were effectively equivalent to long-polling.

I also think that the semantics in this PR are compatible with a long-polling implementation: specifically, on receipt of a poll request (for an incomplete job) the Helper would not respond at all until the job is complete (resulting in a 200 response) or some timeout occurs (resulting in a 202 response).

But now that we're thinking about this, I do wonder if this could be a source of friction between different implementations: the deployment considerations for immediate-polling are different than those for long-polling, so an implementation optimized for immediate-polling may not play well with an implementation optimized for long-polling.

Reply:
I hope the above clarifies some of the confusion. We are still doing it this process in two steps, but the later step can be optimized as mentioned above without compromising on proposed async model.

Reply:
What textual change are you suggesting? My interpretation of the text in this PR is that an implementation is allowed to implement either frequent-polling or long-polling. There are some operational considerations which are not specified (e.g. if long-polling, how long does the Helper wait before giving up & returning 202?) but I think they can be left implementation-dependent.

N.B. as it stands, the initial request must always return a 201 Created -- i.e. the old "sync" model has been replaced by the new "async" model. The prior discussion around this is in #557.

Reply:
> My interpretation of the text in this PR is that an implementation is allowed to implement either frequent-polling or long-polling.

True, with caveats. The GET request that follow PUT/POST requests could long poll. But note that this PR changeset would have inserted a new request that didn't exist before, i.e. long polling on the PUT/POST requests is no longer possible.

But as @suman-ganta states, we can make it so that:
> aggregate init request can return either 200 with proper response body (sync) or 202 without body (async). This makes it backward compatible and supports sync model.

In effect we're raising the question again we thought about in https://github.com/ietf-wg-ppm/draft-ietf-ppm-dap/issues/557#issuecomment-2166938611. I do see value in letting the polling behavior (long or short) be decided amongst implementations, but it does seem to come at the cost of spec complexity.

Reply:
I think we should be careful not to conflate the discussion of:

* Sync vs async aggregation job handling (and whether to support both of these, or only one), first opened in #557 and continued in this PR.
* Frequent-polling vs long-polling, first discussed in this PR.

FWIW/IMO, I am concerned about the need for an extra request per aggregation job implied by the async model, but not enough to accept the complexity of specifying & implementing both sync & async behaviors -- I think we should specify only one of sync & async. I am still thinking through whether it is wise to allow both frequent-polling & long-polling behaviors.

Reply:
I think the PUT should always return 202. I like @suman-ganta's suggestion for the GET. I understand it to be:

> If the response is ready, then return 200. Otherwise, if the response is not ready, the Helper MAY return 202 immediately, or it MAY wait until the response is ready and respond with 200.

Reply:
A textual change may make this clearer.

Reply:
> I think the PUT should always return 202.

Clarify? PUT should always return 201 Created?

> I like @suman-ganta's suggestion for the GET.

I'm fine with this change as well.

> A textual change may make this clearer.

Agreed, I think the behavior is permissible with the text as-is, but it's not obvious.